smarts/core/sumo_road_network.py

```diff
@@ -269,6 +272,58 @@ def _snap_internal_edges(self, lane_to_poly, snap_threshold=2):
                 lane_shape = Polygon(snap(lane_shape, outgoing_shape, snap_threshold))
                 lane_to_poly[lane_id] = lane_shape
 
+    def _snap_internal_holes(self, lane_to_poly, snap_threshold=2):
+        for lane_id in lane_to_poly:
+            lane = self.lane_by_id(lane_id)
+
+            # Only do snapping for internal edge lane holes
+            if not lane.getEdge().isSpecial():
+                continue
+            lane_shape = lane_to_poly[lane_id]
+            for x, y in lane_shape.exterior.coords:
+                nl = self.nearest_lane(
+                    (x, y),
+                    max(10, 2 * self._default_lane_width),
+                    include_junctions=False,
+                )
+                if nl:
+                    nl_shape = lane_to_poly.get(nl.getID())
+                    lane_shape = Polygon(snap(lane_shape, nl_shape, snap_threshold))
+                    lane_to_poly[lane_id] = lane_shape
+
+    def _snap_external_holes(self, lane_to_poly, snap_threshold=2):
+        for lane_id in lane_to_poly:
+            lane = self.lane_by_id(lane_id)
+
+            # Only do snapping for external edge lane holes
+            if lane.getEdge().isSpecial():
+                continue
+
+            incoming = self.lane_by_id(lane_id).getIncoming()
+            if incoming:
+                if incoming[0].getEdge().isSpecial():
+                    continue
+
+            outgoing = self.lane_by_id(lane_id).getOutgoing()
+            if outgoing:
+                outgoing_lane = outgoing[0].getToLane()
+                if outgoing_lane.getEdge().isSpecial():
+                    continue
+
+            lane_shape = lane_to_poly[lane_id]
+            for x, y in lane_shape.exterior.coords:
+                nl = self.nearest_lane(
+                    (x, y),
+                    max(10, 2 * self._default_lane_width),
+                    include_junctions=False,
+                )
+                if nl:
+                    if nl.getEdge().isSpecial():
+                        continue
```

Comment:
I would prefer flattening the above 3 lines to:
```python
    if nl and not nl.getEdge().isSpecial():
```

Also, was a similar check like this not needed in the internal version of this above?

Reply:
No, seems like the internal edges can be snapped with any nearby edges to ensure they are seamed tightly. When I add a similar checker for them above, I still get the same result. So adding this check there seems redundant to me.

Though I have to add a checker for external edges, otherwise it seems to cut out external lanes out of the map if they get snapped to internal lanes. 
This is because if the holes are between an internal and a external lane, then only the internal lane needs to be snapped to the nearby lanes to remove that hole. If we snap the external lanes to internal lanes, the external lane itself will be removed from the map. While if the two adjacent lanes are external (i.e. there are only two lanes), then both lanes need to be snapped to each other to remove the holes completely.